pytorch_lightning/trainer/training_loop.py

```diff
@@ -230,23 +231,30 @@ def _should_add_batch_output_to_epoch_output(self) -> bool:
 
         return False
 
-    def get_optimizers_iterable(self, batch_idx=None):
+    def _should_skip_optimizer(self, opt_idx: int, batch_idx: Optional[int] = None) -> bool:
+        """ Determine if the optimizer should be skipped based on desired frequencies. """
+        active_indices = [idx for (idx, _) in self.get_active_optimizers(batch_idx)]
+        return opt_idx not in active_indices
+
+    def get_active_optimizers(self, batch_idx: Optional[int] = None) -> List[Tuple[int, Optimizer]]:
         """
-        Generates an iterable with (idx, optimizer) for each optimizer.
+        Returns the currently active optimizers. When multiple optimizers are used with different frequencies,
+        only one of the optimizers is active at a time.
+
+        Returns:
+            A list of tuples (opt_idx, optimizer) of currently active optimizers.
         """
         if not self.trainer.optimizer_frequencies:
             # call training_step once per optimizer
             return list(enumerate(self.trainer.optimizers))
 
-        if batch_idx is None:
-            batch_idx = self.total_batch_idx
-
+        batch_idx = self.total_batch_idx if batch_idx is None else batch_idx
```

Comment:
```suggestion
        batch_idx = batch_idx or self.total_batch_idx
```

Reply:
batch_idx can be 0 too. For sanity I recommend not omitting the None check.